test/e2e/framework/autoscaling/autoscaling_utils.go

```diff
@@ -520,15 +521,9 @@ func (rc *ResourceConsumer) EnsureDesiredReplicasInRange(ctx context.Context, mi
 		} else if replicas > maxDesiredReplicas {
 			return false, fmt.Errorf("number of replicas above target")
 		} else {
-			return false, nil // Expected number of replicas found. Continue polling until timeout.
+			return true, nil // Expected number of replicas found. Continue polling until timeout.
 		}
-	})
-	// The call above always returns an error, but if it is timeout, it's OK (condition satisfied all the time).
-	if wait.Interrupted(err) {
-		framework.Logf("Number of replicas was stable over %v", duration)
-		return
-	}
-	framework.ExpectNoErrorWithOffset(1, err)
+	})).WithTimeout(duration).WithPolling(interval).Should(gomega.BeTrue())
```

Comment:
When using `gomega.Consistently`, you are making a statement about some observable fact. The callback should only deliver that fact, the statement then should be (no pun intended) in the `Should`. That way you get all the advantages listed in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts.

What that means here is:
- don't log inside the callback
- return a struct that contains the information that needs to be logged in case of a failure
- check that the state in that struct passes with a matcher; a custom matcher might be the easiest, but you can also compose one (`gomega.HaveField + gomega.BeNumerically`)

You also need to check error handling. Does `rc.GetReplicas` fail on errors? It would be better to return an error and let `Consistently` and `HandleRetry` deal with it.

Regarding `EnsureDesiredReplicasInRange`: that's inside a helper package. When it fails, we want to see where in the test it was used, not a failure inside the helper. See "The E2E framework therefore uses a different approach" in linked section.

Reply:
> You also need to check error handling. Does `rc.GetReplicas` fail on errors? It would be better to return an error and let `Consistently` and `HandleRetry` deal with it.

It doesn't look like it does.  It fails hard with in its helper function after making the API call. If I understand correctly that isn't following the pattern suggested here iether. 
 https://github.com/kubernetes/kubernetes/blob/1bb2773153d5d7395d445feb23171ad8b84afb13/test/e2e/framework/autoscaling/autoscaling_utils.go#L444-L452

Cleaning that up seems like a larger refactor, Is this something that can be a follow up? I am warry of making large changes before code freeze but would like to see these tests passing.

Reply:
I'm okay with fixing this in a follow-up. But please create an issue and link to it in the code.

Reply:
I've updated based on your feedback (great docs btw).  I am not sure if the changes are what you had in mind but I think it is a bit closer to the approach suggested

Reply:
https://github.com/kubernetes/kubernetes/issues/121892

Reply:
```go
// GetObject takes a get function like clientset.CoreV1().Pods(ns).Get
// and the parameters for it and returns a function that executes that get
// operation in a [gomega.Eventually] or [gomega.Consistently].
//
// Delays and retries are handled by [HandleRetry]. A "not found" error is
// a fatal error that causes polling to stop immediately. If that is not
// desired, then wrap the result with [IgnoreNotFound].
func GetObject[T any](get APIGetFunc[T], name string, getOptions metav1.GetOptions) GetFunc[T] {
	return HandleRetry(func(ctx context.Context) (T, error) {
		return get(ctx, name, getOptions)
	})
}
```

@pohly @jsturtevant How about this? It just need do some small changes in GetReplicas func.

```go
// GetReplicas get the replicas
func (rc *ResourceConsumer) GetReplicas(ctx context.Context) int {
	switch rc.kind {
	case KindRC:
		replicationController, err := framework.GetObject(rc.clientSet.CoreV1().ReplicationControllers(rc.nsName).Get, rc.name, metav1.GetOptions{})(ctx)
		framework.ExpectNoError(err)
		if replicationController == nil {
			framework.Failf(rcIsNil)
		}
		return int(replicationController.Status.ReadyReplicas)
```